certora/D3MHub.spec

```diff
@@ -0,0 +1,1376 @@
+    assert(revert22 => lastReverted, "revert22 failed");
+    assert(revert23 => lastReverted, "revert23 failed");
+    assert(revert24 => lastReverted, "revert24 failed");
+    assert(revert25 => lastReverted, "revert25 failed");
+    assert(revert26 => lastReverted, "revert26 failed");
+    assert(revert27 => lastReverted, "revert27 failed");
+
+    assert(lastReverted => revert1  || revert2  || revert3  ||
+                           revert4  || revert5  || revert6  ||
+                           revert7  || revert8  || revert9  ||
+                           revert10 || revert11 || revert12 ||
+                           revert13 || revert14 || revert15 ||
+                           revert16 || revert17 || revert18 ||
+                           revert19 || revert20 || revert21 ||
+                           revert22 || revert23 || revert24 ||
+                           revert25 || revert26 || revert27, "Revert rules are not covering all the cases");
+}
+
+rule exec_exec(bytes32 ilk) {
+    env e;
+
+    require(vat() == vat);
+    require(daiJoin() == daiJoin);
+    require(plan(ilk) == plan);
+    require(pool(ilk) == pool);
+    require(daiJoin.dai() == dai);
+    require(daiJoin.vat() == vat);
+    require(plan.dai() == dai);
+    require(pool.hub() == currentContract);
+    require(pool.vat() == vat);
+    require(pool.dai() == dai);
+
+    uint256 maxDeposit = pool.maxDeposit(e);
+    uint256 assetsBefore = pool.assetBalance(e);
+    uint256 targetAssets = plan.getTargetAssets(e, assetsBefore);
+
+    require(maxDeposit > targetAssets - assetsBefore);
+
+    exec(e, ilk);
+
+    uint256 assetsAfter1 = pool.assetBalance(e);
+
+    uint256 inkAfter1;
+    uint256 artAfter1;
+    inkAfter1, artAfter1 = vat.urns(ilk, pool);
+
+    exec(e, ilk);
+
+    uint256 assetsAfter2 = pool.assetBalance(e);
+
+    uint256 inkAfter2;
+    uint256 artAfter2;
+    inkAfter2, artAfter2 = vat.urns(ilk, pool);
+
+    assert(assetsAfter2 == assetsAfter1, "assetsAfter did not remain as expected");
+    assert(inkAfter2 == inkAfter1, "inkAfter did not remain as expected");
+    assert(artAfter2 == artAfter1, "artAfter did not remain as expected");
+}
+
+rule exit(bytes32 ilk, address usr, uint256 wad) {
+    env e;
+
+    require(vat() == vat);
+    require(pool(ilk) == pool);
+    require(pool.hub() == currentContract);
+    require(pool.vat() == vat);
+    require(pool.share() == share);
+
+    uint256 vatGemSenderBefore = vat.gem(ilk, e.msg.sender);
+    uint256 poolShareUsrBefore = share.balanceOf(usr);
+
+    exit(e, ilk, usr, wad);
+
+    uint256 vatGemSenderAfter = vat.gem(ilk, e.msg.sender);
+    uint256 poolShareUsrAfter = share.balanceOf(usr);
+
+    assert(vatGemSenderAfter == vatGemSenderBefore - wad, "vatGemSender did not decrease by wad amount");
+    assert(usr != pool => poolShareUsrAfter == poolShareUsrBefore + wad, "poolShareUsr did not increase by wad amount");
+}
+
+rule exit_revert(bytes32 ilk, address usr, uint256 wad) {
```

Comment:
does this cover the case where `wad = 0`? We were thinking that we wanted some protection for this function receiving `wad = 0` for these and future pools?

Reply:
It includes the value of `wad = 0` basically `wad` can be any `uint256` value.
However the whole suite is working under the use of `D3MTestPool`, so a different `pool` might generate a different behaviour. We might want to think on adding a "MaliciousPool" where we can prove how the `hub` reacts.